src/connection.js

```diff
@@ -1274,6 +1306,95 @@ class Connection extends EventEmitter {
     this.transitionTo(this.STATE.SENT_LOGIN7_WITH_STANDARD_LOGIN);
   }
 
+  getRequestOptions() {
+    var tokenHost = '169.254.169.254';
+    var tokenEndpoint = '/metadata/identity/oauth2/token';
+    var clinetIdKeyName = '&client_id=';
+    const apiVersionKeyName = '&api-version=';
+    const clinetId = this.config.authentication.options.MSIclientID;
+    const msiEndpoint = process.env.MSI_ENDPOINT;
+    const msiSecret = process.env.MSI_SECRET;
+    var tokenPath = '';
+    var tokenPort = '';
+    var tokenHeaders = { 'Metadata': true };
+    // Api version supported on Azure VM
+    var apiVersion = '2018-02-01';
+
+    // If MSI_ENDPOINT and MSI_SECRET are defined, used these instead of the hard-coded host
+    if (undefined !== msiEndpoint && undefined !== msiSecret) {
+      // Parse the environment variable as a url and set corresponding request components
+      var appUrl = url.parse(msiEndpoint, true);
```

Comment:
What should happen if the URL in the `MSI_ENDPOINT` is misformed? 🤔

Reply:
I could have a check here for that, but these two ENV should come with the Function APP. The process for these two to show up should be automatic. At least from my current understanding for the Azure.

Reply:
I have added a try-catch block for the parse function.  Just wondering, is it ok that we expose a detail error message here? I saw some other method indeed just return a pre-set error message like "Security token could not be authenticated or authorized. "

Reply:
Check the documentation for url.parse, it supposes throw two kinds of error: URI and Type. I added a type check for Type error, so I can represent a corresponding error message. For other errors, I will just return what error message the error provided. 
In terms of the URI, I tried a few different cases, and parse function never throw this type of error.  I also checked the parsing code, and I did not find a place that throws this error. Maybe, it will be thrown in a newer version?